blog/_posts/2021-02-16-improving-scala-library-ecosystem.md

```diff
@@ -0,0 +1,127 @@
+  "org.tpolecat"      %% "doobie-postgres-circe" % "0.10.0"
+)
+~~~
+
+Note that both libraries use Circe for working with JSON.
+
+sbt loads the project build, the project compiles successfully, but when you try to run it, you may get a `NoSuchMethodError` exception at run-time!
+
+What happens here is that one of the transitive dependencies of `akka-http-circe` depends on `circe-core` version 0.11.1, and another transitive dependency of `doobie-postgres-circe` depends on `circe-core` version 0.13.0. Here's an excerpt of the output of sbt’s `dependencyTree` task:
+
+~~~ text
+[info]   +-de.heikoseeberger:akka-http-circe_2.12:1.26.0 [S]
+[info]   | +-io.circe:circe-core_2.12:0.11.1 (evicted by: 0.13.0)
+[info]   |     
+[info]   +-org.tpolecat:doobie-postgres-circe_2.12:0.10.0 [S]
+[info]     +-io.circe:circe-core_2.12:0.13.0 [S]
+~~~
+
+Unfortunately, these two versions of Circe are not binary compatible. Such version conflicts are inevitable in a dependency graph of any practical size. The problem is that we carefully program our code using a statically checked type system, but when it comes to production code we accept swapping out the JAR file with something that our dependency resolver like Coursier and Apache Ivy selected on a whim.
+
+Scala Center, Alexandre Archambault (author of Coursier), and I (Eugene Yokota) have been working on a solution to improve this situation so we can be confident about creating libraries, and using them.
+
+In the next section, I will explain the mechanism that was in place so far in sbt to address this issue, and I will discuss its limits. Then, I will introduce a new solution, which concerns both library authors, and library users.
+
+## Eviction warnings
+
+Actually this is not the first time I've thought of this issue. In 2014, I added [eviction warning][1] feature to sbt 0.13.6. In the lingo of dependency resolver, when you have two candidate versions 0.11.1 and 0.13.0, and when it picks 0.13.0, 0.11.1 is said to be evicted.
+
+~~~ text
+sbt:killer-app> evicted
+[warn] Found version conflict(s) in library dependencies; some are suspected to be binary incompatible:
+[warn]  * io.circe:circe-core_2.12:0.13.0 is selected over 0.11.1
+[warn]      +- org.tpolecat:doobie-postgres-circe_2.12:0.10.0     (depends on 0.13.0)
+[warn]      +- de.heikoseeberger:akka-http-circe_2.12:1.26.0      (depends on 0.11.1)
+[success] Total time: 1 s, completed Feb 1, 2021, 2:59:56 PM
+~~~
+
+Here, we see that sbt successfully detected the conflict for circe-core.
+
+This was a step in the right direction, but it did not work well in reality because I had no way to tell whether two versions of a library would be binary compatible or not. In 2014 what I did was guess that a Java library would adopt [Semantic Versioning][2], and a Scala library would adopt [PVP][3]. (Apparently, using the first two numbers of the version, e.g. 1.2.x, to mean major version has a name, and it's Haskell Package Versioning Policy, or PVP for short).
```

Comment:
I've never heard of PVP in my life. I'm not sure whether introducing it is educational, or simply a distraction. I lean towards distraction. Maybe substitute "epoch.major.minor"? Not sure.

Reply:
At some point we should introduce the term PVP since it's one of the supported version schemes - https://www.scala-sbt.org/1.x/docs/Publishing.html#Version+scheme